runtime-library/utils/type_utils.c

```diff
@@ -283,14 +283,97 @@ get_closure_struct_type(wasm_module_t wasm_module,
     return -1;
 }
 
+static uint32_t
+get_stringref_array_type(wasm_module_t module, wasm_array_type_t *p_array_type_t)
+{
+    uint32_t i, type_count;
+    bool is_mutable = true;
+    type_count = wasm_get_defined_type_count(module);
+    for (i = 0; i < type_count; i++) {
+        wasm_defined_type_t type = wasm_get_defined_type(module, i);
+
+        if (wasm_defined_type_is_array_type(type)) {
+            bool mutable_ref = false;
+            wasm_ref_type_t arr_elem_ref_type = wasm_array_type_get_elem_type(
+                (wasm_array_type_t)type, &mutable_ref);
+
+            if (arr_elem_ref_type.value_type == VALUE_TYPE_STRINGREF
+                && mutable_ref == is_mutable) {
+                if (p_array_type_t) {
+                    *p_array_type_t = (wasm_array_type_t)type;
+                }
+                return i;
+            }
+        }
+    }
+    if (p_array_type_t) {
+        *p_array_type_t = NULL;
+    }
+
+    return -1;
+}
+
 #if WASM_ENABLE_STRINGREF != 0
 wasm_struct_obj_t
 create_wasm_array_with_string(wasm_exec_env_t exec_env, void **ptr,
                               uint32_t arrlen)
 {
-    /* TODO: implement create_wasm_array_with_string based on stringref */
-    bh_assert(false);
-    return NULL;
+    wasm_module_inst_t module_inst = wasm_runtime_get_module_inst(exec_env);
+    wasm_module_t module = wasm_runtime_get_module(module_inst);
+    wasm_local_obj_ref_t local_ref = { 0 };
+    wasm_array_type_t stringref_array_type = NULL;
+
+    uint32_t res_arr_type_idx =
+        get_stringref_array_type(module, &stringref_array_type);
+
+    wasm_struct_type_t res_arr_struct_type = NULL;
+
+    /* get result array struct type */
+    get_array_struct_type(module, res_arr_type_idx, &res_arr_struct_type);
+
+    if (!ptr || !arrlen)
+        return NULL;
+
+    /* create new array */
+    wasm_value_t val = { 0 };
+    val.gc_obj = NULL;
+    wasm_array_obj_t new_arr = wasm_array_obj_new_with_type(
+        exec_env, stringref_array_type, arrlen, &val);
+    /* Push object to local ref to avoid being freed at next allocation */
+    wasm_runtime_push_local_object_ref(exec_env, &local_ref);
+    local_ref.val = (wasm_obj_t)new_arr;
+
+    if (!new_arr) {
+        wasm_runtime_pop_local_object_ref(exec_env);
```

Comment:
Had better call `wasm_runtime_push_local_object_ref` after checking whether `new_arr` is null, then you don't need to pop it here.

Reply:
OK